File: account/src/main/java/serp/project/account/ui/controller/UserController.java
/**
 * Author: QuanTuanHuy
 * Description: Part of Serp Project
 */

package serp.project.account.ui.controller;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.util.Optional;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import serp.project.account.core.domain.constant.Constants;
import serp.project.account.core.domain.dto.request.AssignRoleToUserDto;
import serp.project.account.core.domain.dto.request.GetUserParams;
import serp.project.account.core.usecase.UserUseCase;
import serp.project.account.kernel.utils.AuthUtils;
import serp.project.account.kernel.utils.ResponseUtils;

@RequiredArgsConstructor
@RestController
@RequestMapping("/api/v1/users")
@Slf4j
public class UserController {
    private final UserUseCase userUseCase;

    private final AuthUtils authUtils;
    private final ResponseUtils responseUtil;

    @GetMapping
    public ResponseEntity<?> getUsers(
            @RequestParam(required = false) Integer page,
            @RequestParam(required = false) Integer pageSize,
            @RequestParam(required = false) String sortBy,
            @RequestParam(required = false) String sortDir,
            @RequestParam(required = false) String search,
            @RequestParam(required = false) String status,
            @RequestParam(required = false) Long organizationId) {
        if (!authUtils.canAccessOrganization(organizationId)) {
            organizationId = authUtils.getCurrentTenantId().orElse(null);
        }

        GetUserParams params = GetUserParams.builder()
                .page(page).pageSize(pageSize).sortBy(sortBy).sortDirection(sortDir)
                .search(search)
                .status(status)
                .organizationId(organizationId)
                .build();
        var response = userUseCase.getUsers(params);
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @GetMapping("/profile/me")
    public ResponseEntity<?> getMyProfile() {
        Optional<Long> userIdOpt = authUtils.getCurrentUserId();
        if (userIdOpt.isEmpty()) {
            var response = responseUtil.unauthorized(Constants.ErrorMessage.UNAUTHORIZED);
            return ResponseEntity.status(response.getCode()).body(response);
        }
        var response = userUseCase.getUserProfile(userIdOpt.get());
        return ResponseEntity.status(response.getCode()).body(response);
    }

    @PostMapping("/assign-roles")
    public ResponseEntity<?> assignRoles(@Valid @RequestBody AssignRoleToUserDto request) {
        var response = userUseCase.assignRolesToUser(request);
        return ResponseEntity.status(response.getCode()).body(response);
    }

}

File: api_gateway/src/ui/controller/account/user_controller.go
/*
Author: QuanTuanHuy
Description: Part of Serp Project
*/

package controller

import (
	"github.com/gin-gonic/gin"
	"github.com/serp/api-gateway/src/core/domain/constant"
	request "github.com/serp/api-gateway/src/core/domain/dto/request/account"
	service "github.com/serp/api-gateway/src/core/service/account"
	"github.com/serp/api-gateway/src/kernel/utils"
)

type UserController struct {
	userService service.IUserService
}

func (u *UserController) GetMyProfile(c *gin.Context) {
	res, err := u.userService.GetMyProfile(c.Request.Context())
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralUnauthorized)
		return
	}
	c.JSON(res.Code, res)
}

func (u *UserController) GetUsers(c *gin.Context) {
	page, pageSize, valid := utils.ValidatePaginationParams(c)
	if !valid {
		return
	}

	sortBy := utils.ParseStringQuery(c, "sortBy")
	sortDir := utils.ParseStringQuery(c, "sortDir")
	search := utils.ParseStringQuery(c, "search")
	status := utils.ParseStringQuery(c, "status")
	organizationID := utils.ParseInt64Query(c, "organizationId")

	params := &request.GetUserParams{
		Page:           &page,
		PageSize:       &pageSize,
		SortBy:         sortBy,
		SortDir:        sortDir,
		Search:         search,
		OrganizationID: organizationID,
		Status:         status,
	}

	res, err := u.userService.GetUsers(c.Request.Context(), params)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func (u *UserController) AssignRolesToUser(c *gin.Context) {
	var req request.AssignRoleToUserDto
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.AbortErrorHandle(c, constant.GeneralBadRequest)
		return
	}

	res, err := u.userService.AssignRolesToUser(c.Request.Context(), &req)
	if err != nil {
		utils.AbortErrorHandle(c, constant.GeneralInternalServerError)
		return
	}
	c.JSON(res.Code, res)
}

func NewUserController(userService service.IUserService) *UserController {
	return &UserController{
		userService: userService,
	}
}
